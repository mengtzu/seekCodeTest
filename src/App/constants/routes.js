const root = '/';
const adPacks = '/adPacks';
const checkout = '/checkout';
const confirmation = '/confirmation';
const logout = '/Login/Logout';

export const routes = {
    root,
    adPacks,
    checkout,
    confirmation,
    logout
};

export default routes;
