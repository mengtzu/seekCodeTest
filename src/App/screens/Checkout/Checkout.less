@import (reference) "~seek-style-guide/theme";

.flexRight {
    flex-grow: 1;
    text-align: right;
}

.paymentInfoWrapper {
    @media only screen and (min-width: @responsive-breakpoint) {
        display: flex;
        justify-content: space-between;
    }
}

.lineItemsWrapper {
    @media only screen and (min-width: @responsive-breakpoint) {
        flex-grow: 1;
        margin-right: @grid-gutter-width;
    }
}

.cardInfoWrapper {
    @media only screen and (min-width: @responsive-breakpoint) {
        width: 350px;
    }
}

.lineItemWrapper {
    display: flex;
}

.lineItemPrice {
    .flexRight();
}

.paymentActionWrapper {
    @media only screen and (min-width: @responsive-breakpoint) {
        display: flex;
        justify-content: space-between;
    }
}

.totalWrapper {
    @media only screen and (min-width: @responsive-breakpoint) {
        flex-grow: 1;
        margin-right: @grid-gutter-width;
    }
}

.totalItemWrapper {
    display: flex;
}

.totalItemPrice {
    .flexRight();
}

.cardWrapper {
    display: flex;
}

.creditCardLogo {
    background-image: url('./VISA-logo.png');
    background-size: cover;
    width: 102px;
    height: 54px;
    margin-top: @grid-row-height;
    margin-right: @grid-gutter-width;
}

.creditCardText {
    .flexRight();
}

.actionButtons {
    display: flex;
    justify-content: space-between;

    @media only screen and (min-width: @responsive-breakpoint) {
        width: 350px;
        align-items: flex-end;
    }
}