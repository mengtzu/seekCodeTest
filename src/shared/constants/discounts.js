export const DISCOUNT_TYPE_FLAT_RATE = 'flatRate';
export const DISCOUNT_TYPE_THRESHOLD = 'threshold';
export const DISCOUNT_TYPE_X_FOR_Y = 'xfory';
