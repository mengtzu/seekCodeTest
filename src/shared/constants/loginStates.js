export const AUTH_STATUS_UNAUTHENTICATED = 'pending'; //'unauthenticated' exposes the real SEEK login, don't want that!
export const AUTH_STATUS_AUTHENTICATED = 'authenticated';
