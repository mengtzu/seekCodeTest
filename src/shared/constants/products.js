export const CLASSIC = 'classic';
export const STANDOUT = 'standout';
export const PREMIUM = 'premium';
