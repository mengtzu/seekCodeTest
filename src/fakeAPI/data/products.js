import { CLASSIC, STANDOUT, PREMIUM } from '../../shared/constants/products';

let productsData = {};

//Bracket notation isn't necessarily preferred but here it lets us avoid having to iterate over this object in various places

productsData[CLASSIC] = {
    id: CLASSIC,
    basePrice: 269.99
};

productsData[STANDOUT] = {
    id: STANDOUT,
    basePrice: 322.99
};

productsData[PREMIUM] = {
    id: PREMIUM,
    basePrice: 394.99
}

export default productsData;