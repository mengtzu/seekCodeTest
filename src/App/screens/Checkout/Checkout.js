import React from 'react';
import { connect } from 'react-redux';
import {
    Text,
    PageBlock,
    Card,
    Section,
    Strong,
    Button
} from 'seek-style-guide/react';
import styles from './Checkout.less';
import { formatMoney } from 'accounting-js';
import './VISA-logo.png';

import { CLASSIC, STANDOUT, PREMIUM } from '../../../shared/constants/products';
import { completePayment } from './actions/checkoutActions';
import CodeTestHelpMessage from '../../components/CodeTestHelpMessage/CodeTestHelpMessage';

const mapStateToProps = (state) => {
    return {
        checkout: state.checkout,
        advertiser: state.advertiser
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        handlePayment: (history, checkout) => {
            return completePayment(dispatch, history, checkout);
        }
    }
}

const renderLineItem = (label, lineItem) => {
    if (lineItem.quantity && lineItem.subTotal) {
        return (
          <Card>
              <Section>
                  <Text heading strong>{label}</Text>
                  <div className={styles.lineItemWrapper}>
                      <Text>Quantity: {lineItem.quantity}</Text>
                      <Text className={styles.lineItemPrice} strong>{formatMoney(lineItem.subTotal)}</Text>
                  </div>
              </Section>
          </Card>  
        );
    } else {
        return null;
    }
};

const Checkout = ({ checkout, advertiser, history, handlePayment }) => {

    const onPaymentClick = () => {
        return handlePayment(history, checkout);
    };

    const returnToAdPacks = () => {
        history.push('/adPacks');
    };

    const fourOhFour = () => {
        //Doesn't look right without it but not implemented :)
        history.push('/fourOhFour');
    }

    if (checkout.grandTotal) {
        return (
            <div>
                <PageBlock>
                    <Section header>
                        <Text hero>Checkout</Text>
                    </Section>

                    <div className={styles.paymentInfoWrapper}>
                        <div className={styles.lineItemsWrapper}>
                            {renderLineItem("Classic Ads", checkout[CLASSIC])}
                            {renderLineItem("Standout Ads", checkout[STANDOUT])}
                            {renderLineItem("Premium Ads", checkout[PREMIUM])}
                        </div>
                        <div className={styles.cardInfoWrapper}>
                            <Card>
                                <Section>
                                    <Text>Using saved card for <Strong>{advertiser.displayName}</Strong></Text>
                                    <div className={styles.cardWrapper}>
                                        <div className={styles.creditCardLogo}></div>
                                        <div className={styles.creditCardText}>
                                            <Text>422******2222</Text>
                                            <Text><Strong>Expiry: </Strong> 12/55</Text>
                                        </div>
                                    </div>
                                </Section>
                                <Section>
                                    <Button color="blue" onClick={fourOhFour}>Change</Button>
                                </Section>
                            </Card>
                        </div>
                    </div>

                    <div className={styles.paymentActionWrapper}>
                        <Section>
                            <Text heading>Total</Text>
                            <div className={styles.totalItemWrapper}>
                                <Text>Sub Total:</Text>
                                <Text className={styles.totalItemPrice} strong>{formatMoney(checkout.subTotal)}</Text>
                            </div>
                            <div className={styles.totalItemWrapper}>
                                <Text>GST:</Text>
                                <Text className={styles.totalItemPrice} strong>{formatMoney(checkout.gst)}</Text>
                            </div>
                            <div className={styles.totalItemWrapper}>
                                <Text strong>Grand Total:</Text>
                                <Text className={styles.totalItemPrice} strong>{formatMoney(checkout.grandTotal)}</Text>
                            </div>
                        </Section>
                        <Section>
                            <Button color="pink" loading={checkout.paying} onClick={onPaymentClick}>Pay Now</Button>
                            <Button className={styles.backButton} color="transparent" onClick={returnToAdPacks}>Go back</Button>
                        </Section>
                    </div>

                </PageBlock>

                <CodeTestHelpMessage icon="chotto">
                    <Text heading>Stripe integration not included</Text>
                    <Text>An actual checkout step requires some care around stored cards, confirming payments and making sure the user knows exactly what is happening with their money.  We're skipping those details here, but it just wouldn't seem right to not put this arguably necessary speedbump in for authentic UX!</Text>
                </CodeTestHelpMessage>
            </div>
        )
    } else {
        returnToAdPacks();  //We got here through a back button or somesuch with an empty cart, that's no good!
        return null;
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(Checkout);