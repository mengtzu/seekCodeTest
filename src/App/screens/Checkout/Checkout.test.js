import React from 'react';
import { Checkout } from './Checkout';
import renderer from 'react-test-renderer';

import { CLASSIC, STANDOUT, PREMIUM } from '../../../shared/constants/products';

const shoppingCart = {
    subTotal: 90,
    gst: 9,
    grandTotal: 99
};
shoppingCart[CLASSIC] = {
    quantity: 0,
    subTotal: 0
};
shoppingCart[STANDOUT] = {
    quantity: 4,
    subTotal: 40
};
shoppingCart[PREMIUM] = {
    quantity: 5,
    subTotal: 50
};

const advertiser = {
    displayName: 'Bentenmaru'
};

it('renders correctly', () => {
    const tree = renderer.create(
        <Checkout advertiser={advertiser} checkout={shoppingCart} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});