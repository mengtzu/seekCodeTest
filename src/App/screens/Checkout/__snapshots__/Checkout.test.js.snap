// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className=""
  >
    <div
      className="content"
    >
      <div
        className="root header"
      >
        <div
          className="root hero"
        >
          <span
            className=""
          >
            Checkout
          </span>
        </div>
      </div>
      <div
        className="paymentInfoWrapper"
      >
        <div
          className="lineItemsWrapper"
        >
          <div
            className="root"
          >
            <div
              className="root"
            >
              <div
                className="root heading"
              >
                <span
                  className="root"
                >
                  Standout Ads
                </span>
              </div>
              <div
                className="lineItemWrapper"
              >
                <div
                  className="root"
                >
                  <span
                    className=""
                  >
                    Quantity: 
                    4
                  </span>
                </div>
                <div
                  className="root lineItemPrice"
                >
                  <span
                    className="root"
                  >
                    $40.00
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            className="root"
          >
            <div
              className="root"
            >
              <div
                className="root heading"
              >
                <span
                  className="root"
                >
                  Premium Ads
                </span>
              </div>
              <div
                className="lineItemWrapper"
              >
                <div
                  className="root"
                >
                  <span
                    className=""
                  >
                    Quantity: 
                    5
                  </span>
                </div>
                <div
                  className="root lineItemPrice"
                >
                  <span
                    className="root"
                  >
                    $50.00
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="cardInfoWrapper"
        >
          <div
            className="root"
          >
            <div
              className="root"
            >
              <div
                className="root"
              >
                <span
                  className=""
                >
                  Using saved card for 
                  <strong
                    className="root"
                  >
                    Bentenmaru
                  </strong>
                </span>
              </div>
              <div
                className="cardWrapper"
              >
                <div
                  className="creditCardLogo"
                />
                <div
                  className="creditCardText"
                >
                  <div
                    className="root"
                  >
                    <span
                      className=""
                    >
                      422******2222
                    </span>
                  </div>
                  <div
                    className="root"
                  >
                    <span
                      className=""
                    >
                      <strong
                        className="root"
                      >
                        Expiry: 
                      </strong>
                       12/55
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="root"
            >
              <button
                className="root root_isBlue"
                disabled={false}
                onClick={[Function]}
              >
                Change
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="paymentActionWrapper"
      >
        <div
          className="totalWrapper root"
        >
          <div
            className="root heading"
          >
            <span
              className=""
            >
              Total
            </span>
          </div>
          <div
            className="totalItemWrapper"
          >
            <div
              className="root"
            >
              <span
                className=""
              >
                Sub Total:
              </span>
            </div>
            <div
              className="root totalItemPrice"
            >
              <span
                className="root"
              >
                $90.00
              </span>
            </div>
          </div>
          <div
            className="totalItemWrapper"
          >
            <div
              className="root"
            >
              <span
                className=""
              >
                GST:
              </span>
            </div>
            <div
              className="root totalItemPrice"
            >
              <span
                className="root"
              >
                $9.00
              </span>
            </div>
          </div>
          <div
            className="totalItemWrapper"
          >
            <div
              className="root"
            >
              <span
                className="root"
              >
                Grand Total:
              </span>
            </div>
            <div
              className="root totalItemPrice"
            >
              <span
                className="root"
              >
                $99.00
              </span>
            </div>
          </div>
        </div>
        <div
          className="actionButtons root"
        >
          <button
            className="root root_isPink"
            disabled={false}
            onClick={[Function]}
          >
            Pay Now
          </button>
          <button
            className="root backButton root_isTransparent"
            disabled={false}
            onClick={[Function]}
          >
            Go back
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="root"
  >
    <div
      className=""
    >
      <div
        className="content"
      >
        <div
          className="chotto root"
        >
          <div
            className="root heading"
          >
            <span
              className=""
            >
              Stripe integration not included
            </span>
          </div>
          <div
            className="root"
          >
            <span
              className=""
            >
              An actual checkout step requires some care around stored cards, confirming payments and making sure the user knows exactly what is happening with their money.  We're skipping those details here, but it just wouldn't seem right to not put this arguably necessary speedbump in for authentic UX!
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
